### ISSUE TITLE
External plugin mechanism

### ISSUE THREAD

**Issue by saschagrunert** [cri-o/cri-o#3418]
I was playing around with the new `Sandbox` and `Container` interface in `pkg` with having the knowledge of https://github.com/cri-o/cri-o/issues/2255 in mind. So I did a little test and think we could build-up a plugin system on top of both interfaces.

I prepared  a test repo (https://github.com/saschagrunert/plugin), which bulids a shared object (plugin.so via `go build -buildmode=plugin`) from this code:

```go
package main

import (
	"github.com/cri-o/cri-o/pkg/container"
	"github.com/cri-o/cri-o/pkg/sandbox"
)

type mySandbox string

func NewSandbox() sandbox.Sandbox {
	return mySandbox("Hello, world!")
}

func (mySandbox) Create() error                             { return nil }
func (mySandbox) Start() error                              { return nil }
func (mySandbox) Stop() error                               { return nil }
func (mySandbox) Delete() error                             { return nil }
func (mySandbox) AddContainer(container.Container) error    { return nil }
func (mySandbox) RemoveContainer(container.Container) error { return nil }
```

Then CRI-O could consume the plugin via:
```go
package main

import (
	"plugin"

	"github.com/cri-o/cri-o/pkg/sandbox"
	"github.com/pkg/errors"
	"github.com/sirupsen/logrus"
)

const (
	newSandboxSymbol = "NewSandbox"
)

func main() {
	if err := run(); err != nil {
		logrus.Fatalf("unable to %v", err)
	}
}

func run() error {
	// Open the shared object
	p, err := plugin.Open("./plugin.so")
	if err != nil {
		return errors.Wrap(err, "load plugin.so")
	}

	// Lookup the necessary symbols
	symbol, err := p.Lookup(newSandboxSymbol)
	if err != nil {
		return errors.Wrapf(err, "find symbol %s", newSandboxSymbol)
	}

	// Type assert the symbol
	newSandboxFn, ok := symbol.(func() sandbox.Sandbox)
	if !ok {
		return errors.New("type assert sandbox function")
	}

	// Build the sandbox
	sandbox := newSandboxFn()

	// Should print: Hello, world!
	logrus.Infof("Got sandbox: %s", sandbox)
	return nil
}
```

If all necessary RPC methods would only use the new interfaces, then we could easily provide such a functionality. What do you think?

**Comment by haircommander** [comment]
it would be awesome to encapsulate the kata logic within the interface and finally solve https://github.com/cri-o/cri-o/issues/2255 with it

how much runtime overhead would this be? would we load the .so everytime we run the sandbox? would there be a way of loading it at program startup and having a map of possible sandbox types to build from a factory function, reducing this runtime overhead?

**Comment by saschagrunert** [comment]
> how much runtime overhead would this be? would we load the .so everytime we run the sandbox? would there be a way of loading it at program startup and having a map of possible sandbox types to build from a factory function, reducing this runtime overhead?

I think so yes, on CRI-O server startup we could load all runtime plugins at once and select them based on the configuration. This way we would have no overhead during runtime, only on startup.

**Comment by haircommander** [comment]
SGTM this is a great idea. It would also be great to develop the libpod integration out of tree

**Comment by rhatdan** [comment]
I like this idea as well.
@mrunalp @baude @mheon WDYT?

**Comment by harche** [comment]
Someone tried to use go plugins had commented on [this reddit thread](https://www.reddit.com/r/golang/comments/b6h8qq/is_anyone_actually_using_go_plugins/ejkxd2k/).

**Comment by saschagrunert** [comment]
> Someone tried to use go plugins had commented on [this reddit thread](https://www.reddit.com/r/golang/comments/b6h8qq/is_anyone_actually_using_go_plugins/ejkxd2k/).

Hm, I'm wondering what alternatives we would have… :thinking:

**Comment by rhatdan** [comment]
Well that is a real damper on plugins.

**Comment by harche** [comment]
> > Someone tried to use go plugins had commented on [this reddit thread](https://www.reddit.com/r/golang/comments/b6h8qq/is_anyone_actually_using_go_plugins/ejkxd2k/).
> 
> Hm, I'm wondering what alternatives we would have…

Here is how containerd does it, https://github.com/containerd/containerd/blob/master/PLUGINS.md

Check out the section, `External Plugins`. They either execute a plugin binary or communicate with a plugin service, if you will, via gRPC. But this is much loose coupling between the main code and plugin than what's suggested here in this issue.

**Comment by harche** [comment]
Another good example I can quote is containerd's stream API, https://github.com/containerd/containerd/blob/master/docs/stream_processors.md 

Initially we had coded the image encryption/decryption directly in the diff plugin of the containerd. But later they moved functionality doing operations on the layers to streaming API. Because of this the main containerd code doesn't have any logic of image encryption/decryption but rather this library, https://github.com/containerd/imgcrypt makes it happen using streaming API. 

In a nutshell, they just exec the plugin binary in the path.

**Comment by saschagrunert** [comment]
The exchangeable `runtime-handler` is already kind of a binary-based plugin mechanism, so I think we have to be careful to not overlap features here.

I think the question would be: What should CRI-O still do what the plugin does not do? (aka: What should CRI-O do what libpod later would not do?)

Probably:
- serving the CRI gRPC interface
- providing metrics
- providing a sane configuration
- ensuring the correct order of execution, like a container needs a sandbox (but this is already in the responsibly of the kubelet, right?)

Anything else?

**Comment by vrothberg** [comment]
What are the use cases for plugins? Who's the audience? Plugins can quickly yield maintenance issues; RHEL or SLES kernels are not supported with custom kernel modules for that reason.
